### PR TITLE
`Development`: Add script to create large courses locally

### PR DESCRIPTION
**Pull request by az108**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and Context

Recently, new Performance Guidelines were introduced. To help developers adhere to these guidelines, this script allows them to locally create a large course with a significant number of students, student participations, and submissions. This enables developers to test their pull requests (PRs) with large datasets according to their specific needs.

### Description

large_course_main.py allows you to run the script.
If you want to change the amount of students created, you can modify the `students` variable in the config.ini file.
If you want to change the amount of commits each student should perform in the example exercise, you can modify the `commits` variable in the config.ini file.
If you want to change the amount of programming exercises created, you can modify the `exercises` variable in the config.ini file.
This will automatically perform all the necessary steps:

1.  Authenticate as admin.

2.  Create students.

3.  Create a course.

4.  Add users to the course.

5.  Create a programming exercise.

6.  Add participation and commit for each student.

The provided ReadMe can also be used to get further details if necessary.

### Steps for Testing (This PR can only be tested locally)

Prerequisites:
- 1 Developer

1. Checkout to this branch
2. Start the Artemis Server and Client
3. Navigate to "supporting_scripts/course-scripts"
4. Run large_course_main.py
5. Verify if the students and a Course was created on your Local Setup with a Programming Exercise and Sumbissions based on the amount of students you created.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive script for automated course setup, including user creation, course management, and programming exercise integration.
	- Added functions for managing user enrollments and programming exercises within courses.
	- Provided a cleanup option to delete created students after testing.

- **Documentation**
	- Enhanced performance review guidelines in the pull request template, emphasizing the importance of high performance for large course setups.
	- Added detailed README.md files outlining setup instructions, usage, and project purpose for educators and administrators.

### CODE REVIEW

File: supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/deleteStudents.py
Comment (JohannesStoehr) [inline supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/deleteStudents.py:2]: GitHub claims that this is unused

**Review by JohannesStoehr — COMMENTED** [review]
Looks good overall, left two inline comments:

File: supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/deleteStudents.py
Comment (sarpsahinalp) [inline supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/deleteStudents.py:3]: It seems like configparser is unused as well

**Review by sarpsahinalp — COMMENTED** [review]
Apart the from the mentioned unused import code lgtm

**Review by MarkusPaulsen — CHANGES_REQUESTED** [review]
Additionally:
1. Add type hints at least to all functions (parameter and return type).
2. Make the creation of a programming exercise (Step 5) optional, so that we can also use it on an existing programming exercise.

File: supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/createCourse.py
Comment (az108) [reply supporting_scripts/local-course-scripts/large-course-with-programming-exercise-sumbissions-setup/createCourse.py:51]: not supported by .ini file

File: supporting_scripts/course-scripts/quick-course-setup/utils.py
Comment (JohannesWt) [inline supporting_scripts/course-scripts/quick-course-setup/utils.py:28]: just a small idea: you could add `/api` also to the `server_url`

**Review by sarpsahinalp — DISMISSED** [review]
Code lgtm after changes

**Review by JohannesWt — COMMENTED** [review]
Just a small (optional) remark

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:100]: I think `Section 1` and `Section 2` should have the same headline type

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:9]: I think we could display the command right away, I do not think that we need to test the scripts for 2 python versions - you are directly referring to python 3.9 anyways
```suggestion
- Ensure that Python (preferably version 3.9) is installed on your system.
   ```shell
   python3 --version
   ```
```

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:19]: I think we should directly link the config file and push a version with default settings. Why should every developer have to create the config file, think about where to create it and which values need to be present? 
This can be prevented by simply supplying a default configuration
```suggestion
- Configure the [config.ini](./config.ini) file according to your local or test server setup.
```

File: supporting_scripts/course-scripts/quick-course-setup/utils.py
Comment (az108) [reply supporting_scripts/course-scripts/quick-course-setup/utils.py]: did it for client aswell

**Review by florian-glombik — CHANGES_REQUESTED** [review]
I already got some comments, could you please add a template `config.ini` file and then just add it to the `.gitignore`?

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [reply supporting_scripts/course-scripts/quick-course-setup/README.md]: Why do we not directly link the file? 

Why do we need to list variables that can be configured if they can be seen in the linked `config.ini` anyways?

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:9]: Why do we not directly supply the command to check the current version? Not everybody might be fluent in python and know that you have to check with `python3` instead of `python`

```suggestion
- Ensure that Python (preferably version 3.9) is installed on your system.
   ```shell
   python3 --version
   ```
```

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:16]: Lets fix the formatting here
```suggestion
  - Add a Python environment by configuring the Python interpreter.
  
     ![IntelliJ Python Facet Configuration](./images/facets-config.png)
  - Add a module in IntelliJ by navigating to `File > Project Structure > Modules > Add - Python`.

    ![IntelliJ Module Configuration](./images/module-config.png)
```

Comment (florian-glombik) [reply supporting_scripts/course-scripts/quick-course-setup/README.md:16]: Could you also please update the screenshots? 
For me it looks different by now - how does your configuration look like? Is this only due to some local changes on my setup?

image
image

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:21]: I think this might be confusing otherwise
```suggestion
- If packages are missing when running the script, install the necessary Python packages using the following command (replace `<packageName>` with the actual package name and the python version with your used python version):
```

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:5]: Why is `setup` on a lower headline level than `usage`?

File: supporting_scripts/course-scripts/quick-course-setup/create_course.py
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/create_course.py:94]: I think we could improve the error message here, I ran into the problem while testing 😅 
```suggestion
            f"Double check whether the courseShortName {course_short_name} is valid (e.g. no special characters such as '-') not already used for another course!\n"
```

**Review by JohannesWt — DISMISSED** [review]
Code 👍

**Review by florian-glombik — CHANGES_REQUESTED** [review]
For me no logs are displayed at all when running the course setup scripts (create users, authenticate users) which seems to be due to a missing configuration of the logger

Maybe the following could work?
Creating a file `logging_config.py`with the content
```
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
```

and then instead of directly importing logging in each file import `from logging_config import logging`

There might be other better solutions, I am not sure who has an expertise in python in the team (at leat not me) 😅

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [reply supporting_scripts/course-scripts/quick-course-setup/README.md:16]: Without the empty line and the spacing at the start the pictures are not aligned properly for me 
image

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:122]: Could we directly add the option to execute the script here? :) 
```suggestion
6. You can use the play button within IntelliJ (if Python is configured properly) to run the script. (Run large_course_main.py)
   ```shell
    python3 large_course_main.py
    ```
```

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:138]: Could we please add a flag with which the command can be run instead of manual adjustments in the script?

**Review by florian-glombik — CHANGES_REQUESTED** [review]
Thanks for fixing the logging and updating the ReadMe!

The course setup script still works and I have tested the script with 10 students and 2 submissions each, worked like a charm when executed the first time! 

### Further change requests
I unfortunately do have some further requests:

* Could you please mention this script in the pull request template at the respective steps in the performance review section? This will help new devs to actually use the work you have put in this PR! 
* If I run the script twice with the same configuration I get the following output and error because the course does already exist:
   ```
   2024-09-03 21:02:02,336 - INFO - Authentication successful for user artemis_admin
    2024-09-03 21:02:02,343 - INFO - User student1 already exists.
    2024-09-03 21:02:02,349 - INFO - User student2 already exists.
    2024-09-03 21:02:02,355 - INFO - User student3 already exists.
    2024-09-03 21:02:02,368 - INFO - User student4 already exists.
    2024-09-03 21:02:02,377 - INFO - User student5 already exists.
    2024-09-03 21:02:02,384 - INFO - User student6 already exists.
    2024-09-03 21:02:02,391 - INFO - User student7 already exists.
    2024-09-03 21:02:02,398 - INFO - User student8 already exists.
    2024-09-03 21:02:02,405 - INFO - User student9 already exists.
    2024-09-03 21:02:02,415 - INFO - User student10 already exists.
    2024-09-03 21:02:02,422 - ERROR - Problem with the group 'students' and interacting with a test server? Is 'is_local_course' in 'config.ini' set to 'False'?
    Traceback (most recent call last):
    File "/Users/florianglombik/IdeaProjects/Artemis-2024-08-05/supporting_scripts/course-scripts/quick-course-setup/large_course_main.py", line 72, in <module>
    main()
    File "/Users/florianglombik/IdeaProjects/Artemis-2024-08-05/supporting_scripts/course-scripts/quick-course-setup/large_course_main.py", line 43, in main
    created_course_response = create_course(session)
    File "/Users/florianglombik/IdeaProjects/Artemis-2024-08-05/supporting_scripts/course-scripts/quick-course-setup/create_course.py", line 92, in create_course
    raise Exception(
    Exception: Could not create course NewCourseWithManyExercises; Status code: 400
    Double check whether the courseShortName NewCourseWithManyExercises is valid (e.g. no special characters such as '-') not already used for another course!
    Response content:
   ```
    Would it be possible to check if this is a real error or if the course already exists and if the course already exists just keep going with the script?
* Would it be possible to read the name of the exercise from the config file aswell? So it would be possible to create multiple exercises (maybe one for 200, 1000, 2000, 5000 students). The default value could just be the current value.

File: supporting_scripts/course-scripts/quick-course-setup/large_course_main.py
Comment (pzdr7) [inline supporting_scripts/course-scripts/quick-course-setup/large_course_main.py:43]: When running the script locally and using it twice, I got the error

```
2024-09-04 07:07:13,719 - ERROR - Problem with the group 'students' and interacting with a test server? Is 'is_local_course' in 'config.ini' set to 'False'?
Exception: Could not create course Local Course; Status code: 400
Double check whether the courseShortName LocalCourse is valid (e.g. no special characters such as '-') not already used for another course!
```

The workaround is to manually delete the course. Is there any way to remove the course apart from that?

**Review by pzdr7 — CHANGES_REQUESTED** [review]

**Review by florian-glombik — CHANGES_REQUESTED** [review]
* Thanks for adding the option to run it on an existing course and supplying a useful error message! 
```
(venv) florianglombik@MBPvonForian181 quick-course-setup % python3 large_course_main.py
2024-09-04 17:00:16,529 - INFO - Authentication successful for user artemis_admin
2024-09-04 17:00:16,537 - INFO - User student1 already exists.
2024-09-04 17:00:16,545 - INFO - User student2 already exists.
2024-09-04 17:00:16,553 - INFO - Course with shortName LocalCourse already exists. Please provide the course ID in the config file and set create_course to FALSE.
```
### Change Requests
* With the latest changes the course creation script does not work anymore for me
```
(venv) florianglombik@MBPvonForian181 quick-course-setup % python3 create_course.py
2024-09-04 16:58:34,282 - INFO - Authentication successful for user artemis_admin
2024-09-04 16:58:34,317 - INFO - Created course Local Course 123 with shortName LocalCourse123 
 {'id': 413, 'title': 'Local Course 123', 'shortName': 'LocalCourse123', 'studentGroupName': 'students', 'teachingAssistantGroupName': 'tutors', 'editorGroupName': 'editors', 'instructorGroupName': 'instructors', 'testCourse': False, 'onlineCourse': False, 'courseInformationSharingConfiguration': 'COMMUNICATION_AND_MESSAGING', 'maxComplaints': 3, 'maxTeamComplaints': 3, 'maxComplaintTimeDays': 7, 'maxRequestMoreFeedbackTimeDays': 7, 'maxComplaintTextLimit': 2000, 'maxComplaintResponseTextLimit': 2000, 'enrollmentEnabled': False, 'unenrollmentEnabled': False, 'accuracyOfScores': 1, 'restrictedAthenaModulesAccess': False, 'learningPathsEnabled': False, 'studentCourseAnalyticsDashboardEnabled': False, 'requestMoreFeedbackEnabled': True, 'complaintsEnabled': True}
Traceback (most recent call last):
  File "/Users/florianglombik/IdeaProjects/Artemis-2024-08-05/supporting_scripts/course-scripts/quick-course-setup/create_course.py", line 115, in <module>
    main()
  File "/Users/florianglombik/IdeaProjects/Artemis-2024-08-05/supporting_scripts/course-scripts/quick-course-setup/create_course.py", line 109, in main
    response_data = created_course_response.json()
AttributeError: 'dict' object has no attribute 'json'
```

### Follow Up
While we can now create a huge amount of submissions with this script I am not sure about the usefulness of the create submissions as they just seem to trigger a commit from the code editor without any actual changes. All students will just have a score of 0. 
Could you please verify with your advisor that this is intended and sufficient for the script? 

From my point of view this does not need to block this PR as we can build up on this PR to create submissions with actual content.

File: supporting_scripts/course-scripts/quick-course-setup/manage_programming_exercise.py
Comment (pzdr7) [inline supporting_scripts/course-scripts/quick-course-setup/manage_programming_exercise.py:3]: Do you use `List` here at all?

```suggestion
from typing import Dict, Any
```

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (pzdr7) [reply supporting_scripts/course-scripts/quick-course-setup/README.md]: Unless you disable user creation, Playwright will automatically create the accounts for you: https://docs.artemis.cit.tum.de/dev/playwright.html

So the Cypress test users are no longer required here.

Comment (az108) [reply supporting_scripts/course-scripts/quick-course-setup/README.md]: Removed based on @pzdr7 comment that we do not use this anymore

**Comment by az108** [comment]
@florian-glombik Yes the fact that the sumbissions are empty commits is wanted for the script (discussed with Stephan and Markus). Extending it to different commits should not be too difficult in future PRs as I tried to provide all necessary elements in the script already.

File: supporting_scripts/course-scripts/quick-course-setup/create_users.py
Comment (SimonEntholzer) [inline supporting_scripts/course-scripts/quick-course-setup/create_users.py:49]: As in my experience it took quite some time to create users (when creating a lot, like ~1000),  would it maybe make sense to concurrently send create user requests from multiple threads?

File: supporting_scripts/course-scripts/quick-course-setup/README.md
Comment (SimonEntholzer) [inline supporting_scripts/course-scripts/quick-course-setup/README.md:30]: In my opinion it would be advisable to use a virtual environment when working with python, maybe you could add that to the docs here.

https://new.pythonforengineers.com/blog/python-tip-always-use-a-virtual-environment/

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Really nice and helpful scripts :+1: I added two small suggestions

File: supporting_scripts/course-scripts/quick-course-setup/config.ini
Comment (florian-glombik) [inline supporting_scripts/course-scripts/quick-course-setup/config.ini:4]: I think this default might be more useful? I do not have a strong opinion, does not need to be changed for me :) 

```suggestion
students = 2000
commits = 2
exercises = 1
```

**Review by florian-glombik — DISMISSED** [review]
Tested locally, all scripts work for me and the description seems clear and understandable to me

Thank you for taking the time for the adjustments! 🙏 

### What I tested
* setup as described
* executing all scripts
    * course creation scripts work
    * large course participation generation works
    * deletion of users from participation generation does not interfere with other course settings

**Review by florian-glombik — DISMISSED** [review]
Re-approve (test) after minor configuration changes since  https://github.com/ls1intum/Artemis/pull/9252#pullrequestreview-2281543439

**Review by florian-glombik — APPROVED** [review]
Reapprove after minor changes since https://github.com/ls1intum/Artemis/pull/9252#pullrequestreview-2282436575

**Review by SimonEntholzer — APPROVED** [review]
Code lgtm, thanks for adding the suggestions :+1:

**Review by pzdr7 — APPROVED** [review]
Thanks for implementing the code changes 👍

**Review by JohannesWt — APPROVED** [review]
Reapprove Code 👍

**Comment by krusche** [comment]
Great addition, thank you. For a follow-up: can we also allow to create large courses on test servers?

**Pull request merged**
